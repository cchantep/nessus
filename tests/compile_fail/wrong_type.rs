// This test verifies that using the wrong type will cause a compile error

use nessus::{FieldName, FieldWitnesses, HasField};
use serde::{Serialize, Deserialize};

#[derive(Debug, Clone, Serialize, Deserialize, FieldWitnesses)]
struct Product {
    Price: f64,
}

pub fn test_wrong_type<F, T, V>(value: V) -> (&'static str, String) 
where
    F: FieldName,
    T: HasField<F, Value = V>,
    V: ToString
{
    (F::field_name(), value.to_string())
}

fn main() {
    // This should fail to compile because we're trying to use a string for a f64 field
    let price_filter = test_wrong_type::<product_fields::Price, Product, _>("wrong type".to_string());
}
