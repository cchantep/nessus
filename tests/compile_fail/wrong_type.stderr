error[E0308]: mismatched types
  --> tests/compile_fail/wrong_type.rs:21:77
   |
21 |     let price_filter = test_wrong_type::<product_fields::Price, Product, _>("wrong type".to_string());
   |                        ---------------------------------------------------- ^^^^^^^^^^^^^^^^^^^^^^^^ expected `f64`, found `String`
   |                        |
   |                        arguments to this function are incorrect
   |
note: function defined here
  --> tests/compile_fail/wrong_type.rs:10:8
   |
10 | pub fn test_wrong_type<F, T, V>(value: V) -> (&'static str, String)
   |        ^^^^^^^^^^^^^^^          --------
